### DISCUSSION TITLE
Adding Geolocations(5 of) and sending data via SMS to a user

### DISCUSSION THREAD

**Discussion by Darksplat** [microchip-pic-avr-solutions/avr-iot-cellular-arduino-library#24] [Ideas]
Hi I mentioned this on your YouTube video about this nice little device and was wondering about integrating geofencing into your code and sending it via an SMS when it was in or out of that geoloction. Would the code be something like this?

`
#include <Adafruit_GPS.h>
#include <Arduino.h>
#include <http_client.h>
#include <led_ctrl.h>
#include <log.h>
#include <low_power.h>
#include <lte.h>
#include <mcp9808.h>
#include <veml3328.h>
#include <TinyGPS++.h> // Include the TinyGPS++ library for distance calculation
#include <GSM.h> // Include the library for GSM communication

#define GPSSerial Serial2
#define HTTP_DOMAIN "<your server IP here>"
#define MAX_GEOFENCES 5

// Define the user's phone number
const char* userPhoneNumber = "xxxxxxxxxxx";

// Define a struct to hold geofence information
struct Geofence {
    String name; // Name of the geofence
    float latitude;
    float longitude;
    float radius;
    bool inside; // Flag to indicate if device is inside this geofence
};

// Array to hold geofences
Geofence geofences[MAX_GEOFENCES];

// Number of geofences currently defined
int numGeofences = 0;

// Include the TinyGPS++ library for distance calculation
TinyGPSPlus gps;

// Enum for the state of the device
enum class State { NOT_CONNECTED, CONNECTED, CONNECTED_WITH_FIX };

// Global state variable
static State state = State::NOT_CONNECTED;

// Global variables for GPS data
static char latitude[16] = "0";
static char longitude[16] = "0";
static char time[24] = "0";
static float speed = 0.0;
static int quality = 0;
static int numSatellites = 0;

// Flag to indicate whether GPS data has been parsed
static bool has_parsed = false;

// Last time data was sent to server
unsigned long lastDataSentTime = 0;

GSM gsmAccess; // Initialize GSM module
GSM_SMS sms; // Initialize SMS module

// Function prototypes
void initializeGPS();
void initializeHTTP();
void sendData();
void connectToNetwork();
void parseGPSMessages();
void checkGeofences();
void sendSMS(const char* message);

// Function to initialize GPS and HTTP
void setup() {
    LedCtrl.begin();
    LedCtrl.startupCycle();

    Log.begin(115200);
    Log.info(F("Starting AVR-IoT Cellular Adafruit GPS example"));

    // We configure the low power module for power down configuration, where
    // the LTE modem and the CPU will be powered down
    LowPower.configurePowerDown();

    // Make sure sensors are turned off
    Veml3328.begin();
    Mcp9808.begin();
    Veml3328.shutdown();
    Mcp9808.shutdown();

    initializeGPS();
    initializeHTTP();

    // Example geofences
    // Uncomment and customize to add more geofences
    geofences[0] = { "Melbourne Park", -37.8136, 144.9631, 200, false };
    // geofences[1] = { "Custom Geofence Name", latitude, longitude, radius, false };

    numGeofences = 1; // Update the number of geofences
}

// Main loop
void loop() {
    unsigned long currentMillis = millis();
    if (currentMillis - lastDataSentTime >= 15000) {
        lastDataSentTime = currentMillis;
        parseGPSMessages();
        if (state == State::CONNECTED_WITH_FIX && has_parsed) {
            sendData();
            checkGeofences();
        }
    }
}

// Function to initialize GPS
void initializeGPS() {
    GPSSerial.swap(1);
    GPS.begin(9600);

    // Enable RMC & GGA output messages
    GPS.sendCommand(PMTK_SET_NMEA_OUTPUT_RMCGGA);

    // Set the update rate to 1Hz
    GPS.sendCommand(PMTK_SET_NMEA_UPDATE_1HZ);

    // Disable updates on antenna status
    GPS.sendCommand(PGCMD_NOANTENNA);

    delay(1000);

    // Ask for firmware version
    GPSSerial.println(PMTK_Q_RELEASE);
}

// Function to initialize HTTP client
void initializeHTTP() {
    if (!HttpClient.configure(HTTP_DOMAIN, 8080, false)) {
        Log.info(F("Failed to configure HTTP client"));
    } else {
        Log.info(F("Configured HTTP"));
    }
}

// Function to send data to the server
void sendData() {
    char data[120] = "";
    speed = gps.speed.kmph();
    quality = gps.satellites.isValid() ? gps.satellites.value() : 0;
    numSatellites = gps.satellites.isValid() ? gps.satellites.value() : 0;

    sprintf_P(data,
              PSTR("{\"lat\":\"%s\",\"lon\":\"%s\",\"time\":\"%s\",\"speed\":%.2f,\"quality\":%d,\"num_satellites\":%d}"),
              latitude,
              longitude,
              time,
              speed,
              quality,
              numSatellites);

    HttpResponse response = HttpClient.post("/data", data);

    Log.infof(F("POST - status code: %u, data size: %u\r\n"),
              response.status_code,
              response.data_size);

    if (response.status_code != 0) {
        String body = HttpClient.readBody(512);

        if (body != "") {
            Log.infof(F("Response: %s\r\n"), body.c_str());
        }
    }
}

// Function to connect to the network
void connectToNetwork() {
    // If we already are connected, don't do anything
    if (!Lte.isConnected()) {
        while (!Lte.begin()) {}
        Log.infof(F("Connected to operator: %s\r\n"),
                  Lte.getOperator().c_str());
    }

    state = State::CONNECTED;
}

// Function to parse GPS messages
void parseGPSMessages() {
    // Read the incoming messages, needn't do anything with them yet as that is
    // taken care of by the newNMEAReceived() function.
    if (GPS.available()) {
        GPS.read();
    }

    if (GPS.newNMEAreceived()) {
        if (!GPS.parse(GPS.lastNMEA())) {
            // If we fail to parse the NMEA, wait for the next one
            return;
        } else {
            Log.rawf(F("Timestamp: %d/%d/20%d %d:%d:%d GMT+0 \r\n"),
                     GPS.day,
                     GPS.month,
                     GPS.year,
                     GPS.hour,
                     GPS.minute,
                     GPS.seconds);

            Log.rawf(F("Fix: %d, quality: %d\r\n"), GPS.fix, GPS.fixquality);

            if (GPS.fix) {
                // Need to convert all floats to strings
                dtostrf(GPS.latitudeDegrees, 2, 6, latitude);
                dtostrf(GPS.longitudeDegrees, 3, 6, longitude);

                sprintf_P(time,
                          PSTR("%02d/%02d/20%02d %02d:%02d:%02d"),
                          GPS.day,
                          GPS.month,
                          GPS.year,
                          GPS.hour,
                          GPS.minute,
                          GPS.seconds);

                Log.rawf(F("Location: %s N, %s E\r\n"), latitude, longitude);
                Log.rawf(F("Speed: %.2f km/h\r\n"), gps.speed.kmph());
                Log.rawf(F("Satellites: %d\r\n"), gps.satellites.isValid() ? gps.satellites.value() : 0);
                Log.rawf(F("\r\n"));

                has_parsed = true;
            } else {
                Log.info(F("Waiting for GPS fix..."));
            }
        }
    }
}

// Function to check if the device is inside a geofence
void checkGeofences() {
    for (int i = 0; i < numGeofences; ++i) {
        float distance = TinyGPSPlus::distanceBetween(
            gps.location.lat(), gps.location.lng(),
            geofences[i].latitude, geofences[i].longitude);

        bool wasInside = geofences[i].inside; // Previous state

        if (distance <= geofences[i].radius) {
            geofences[i].inside = true;
        } else {
            geofences[i].inside = false;
        }

        // Check if state has changed
        if (wasInside != geofences[i].inside) {
            // Device has moved in or out of this geofence
            String message = "Hi, your GPS tracker device has ";
            if (geofences[i].inside) {
                message += "moved inside";
            } else {
                message += "moved outside";
            }
            message += " your geofence called " + geofences[i].name;
            sendSMS(message.c_str());
        }
    }
}

// Function to send an SMS
void sendSMS(const char* message) {
    // Define APN settings
    char apn[] = "live.vodafone.com"; // APN provided by Vodafone
    char user[] = ""; // Vodafone may not require username
    char password[] = ""; // Vodafone may not require password

/*

| Country   | Carrier      | APN                |
|-----------|--------------|--------------------|
| Australia | Telstra                     | telstra.ipn        |
|                 |                                 | telstra.wap        |
|                 | Vodafone                | live.vodafone.com  |
|           | Optus.                           | connect            |
|           |                                       | yesinternet        |
|           |                                       | connectme          |
|           |  Aldi.                              | mdata.net.au       |
|           |  boost.                           | telstra.mms        |
|           | Moose / Away Sim        | yesinternet        |
| New Zealand | Vodafone     | vodafone.net.nz    |
|             | Spark                       |   internet                 |
|             | 2 Degrees               | internet           |
|             | Kogan                      | live.vodafone.com |
|             | The Warehouse      | warehouse                 |
|             | Skinny                      |  wapaccess.co.nz   |
|             | Slingshot                 | wapaccess.co.nz    |

This Australian/New Zealand table organises the information by country and carrier, listing the corresponding
Access Point Names (APNs). Please note that some carriers may have multiple APNs for
different services (e.g., internet, MMS), and some carriers may have specific requirements
or settings for certain services.

*/

    // Connect to the GSM network
    gsmAccess.begin(apn, user, password);

    // Define recipient's phone number
    char recipient[] = userPhoneNumber;

    // Send SMS message
    sms.beginSMS(recipient);
    sms.print(message);
    sms.endSMS();
}

`

Please feel free to use this and modify it and use it as an example. Also any feedback would be nice on the code corrections etc as I don't have an actual device to test this against as I don't want to spend the money on something That doesn't suit my needs

Cheers
Jeremy

**Comment by M65649** [comment]
Hello @Darksplat,

This is a really cool example. Regarding the functionality to send SMS, the modem has the functionality to do this without the `GSM_SMS` and `GSM` modules. We have not implemented this in the library as of now, but it is fully possible by utilizing AT commands directly. Look at section 14 at the AT command section for the modem here: https://iot.microchip.com/docs/assets/files/Monarch2-LR8.0-ATCommandsReferenceManual-Rev.9-da07a5665e87441e408fd9a9fb47cfc8.pdf. Another maker has made a tutorial on how to send SMS to the AVR-IoT Cellular Mini, you can find it here: https://community.element14.com/products/roadtest/rv/roadtest_reviews/1633/avr-iot_cellular_min.

Sending from the AVR-IoT Cellular Mini is not much more work, and I'm happy to guide you through it should you have any more questions. We have an example to send AT commands to the modem directly here: https://github.com/microchip-pic-avr-solutions/avr-iot-cellular-arduino-library/blob/main/examples/custom_at_commands/custom_at_commands.ino
